File: ios/ReactNativeCameraKit/CodeFormat.swift
//
//  CodeFormat.swift
//  ReactNativeCameraKit
//
//  Created by Imdad on 2023-12-22.
//

import Foundation
import AVFoundation

enum CodeFormat: String, CaseIterable {
    case code128 = "code-128"
    case code39 = "code-39"
    case code39Mod43 = "code-39-mod-43"
    case code93 = "code-93"
    case ean13 = "ean-13"
    case ean8 = "ean-8"
    case itf14 = "itf-14"
    case upce = "upc-e"
    case qr = "qr"
    case pdf417 = "pdf-417"
    case aztec = "aztec"
    case dataMatrix = "data-matrix"
    case interleaved2of5 = "interleaved-2of5"
    case unknown = "unknown"
    @available(iOS 15.4, *)
    case codabar = "codabar"
    @available(iOS 15.4, *)
    case gs1DataBar = "gs1-data-bar"
    @available(iOS 15.4, *)
    case gs1DataBarLimited = "gs1-data-bar-limited"
    @available(iOS 15.4, *)
    case gs1DataBarExpanded = "gs1-data-bar-expanded"

    static var allCases: [CodeFormat] {
        var supportedBarcodeTypes: [CodeFormat] =
        [.upce, .code39, .code39Mod43,
         .ean13, .ean8, .code93, .code128,
         .pdf417, .qr, .itf14, .aztec,
         .dataMatrix, .interleaved2of5]
        
        if #available(iOS 15.4, *) {
            supportedBarcodeTypes.append(contentsOf: [
                .codabar, .gs1DataBar, .gs1DataBarLimited, .gs1DataBarExpanded
            ])
        }
        
        return supportedBarcodeTypes
    }

    // Convert from AVMetadataObject.ObjectType to CodeFormat
    static func fromAVMetadataObjectType(_ type: AVMetadataObject.ObjectType) -> CodeFormat {
        switch type {
        case .code128: return .code128
        case .code39: return .code39
        case .code39Mod43: return .code39Mod43
        case .code93: return .code93
        case .ean13: return .ean13
        case .ean8: return .ean8
        case .itf14: return .itf14
        case .upce: return .upce
        case .qr: return .qr
        case .pdf417: return .pdf417
        case .aztec: return .aztec
        case .dataMatrix: return .dataMatrix
        case .interleaved2of5: return .interleaved2of5
        @available(iOS 15.4, *)
        case .codabar: return .codabar
        @available(iOS 15.4, *)
        case .gs1DataBar: return .gs1DataBar
        @available(iOS 15.4, *)
        case .gs1DataBarLimited: return .gs1DataBarLimited
        @available(iOS 15.4, *)
        case .gs1DataBarExpanded: return .gs1DataBarExpanded
        default: return .unknown
        }
    }

    // Convert from CodeFormat to AVMetadataObject.ObjectType
    func toAVMetadataObjectType() -> AVMetadataObject.ObjectType {
        switch self {
        case .code128: return .code128
        case .code39: return .code39
        case .code39Mod43: return .code39Mod43
        case .code93: return .code93
        case .ean13: return .ean13
        case .ean8: return .ean8
        case .itf14: return .itf14
        case .upce: return .upce
        case .qr: return .qr
        case .pdf417: return .pdf417
        case .aztec: return .aztec
        case .dataMatrix: return .dataMatrix
        case .interleaved2of5: return .interleaved2of5
        @available(iOS 15.4, *)
        case .codabar: return .codabar
        @available(iOS 15.4, *)
        case .gs1DataBar: return .gs1DataBar
        @available(iOS 15.4, *)
        case .gs1DataBarLimited: return .gs1DataBarLimited
        @available(iOS 15.4, *)
        case .gs1DataBarExpanded: return .gs1DataBarExpanded
        case .unknown: return .init(rawValue: "unknown")
        }
    }
}

File: android/src/main/java/com/rncamerakit/CodeFormat.kt
package com.rncamerakit

import com.google.mlkit.vision.barcode.common.Barcode

enum class CodeFormat(val code: String) {
    CODE_128("code-128"),
    CODE_39("code-39"),
    CODE_93("code-93"),
    CODABAR("codabar"),
    EAN_13("ean-13"),
    EAN_8("ean-8"),
    ITF("itf"),
    UPC_E("upc-e"),
    QR("qr"),
    PDF_417("pdf-417"),
    AZTEC("aztec"),
    DATA_MATRIX("data-matrix"),
    UNKNOWN("unknown");

    fun toBarcodeType(): Int {
        return when (this) {
            CODE_128 -> Barcode.FORMAT_CODE_128
            CODE_39 -> Barcode.FORMAT_CODE_39
            CODE_93 -> Barcode.FORMAT_CODE_93
            CODABAR -> Barcode.FORMAT_CODABAR
            EAN_13 -> Barcode.FORMAT_EAN_13
            EAN_8 -> Barcode.FORMAT_EAN_8
            ITF -> Barcode.FORMAT_ITF
            UPC_E -> Barcode.FORMAT_UPC_E
            QR -> Barcode.FORMAT_QR_CODE
            PDF_417 -> Barcode.FORMAT_PDF417
            AZTEC -> Barcode.FORMAT_AZTEC
            DATA_MATRIX -> Barcode.FORMAT_DATA_MATRIX
            UNKNOWN -> -1 // Or any other default value you prefer
        }
    }

    companion object {
        fun fromBarcodeType(@Barcode.BarcodeFormat barcodeType: Int): CodeFormat =
            when (barcodeType) {
                Barcode.FORMAT_CODE_128 -> CODE_128
                Barcode.FORMAT_CODE_39 -> CODE_39
                Barcode.FORMAT_CODE_93 -> CODE_93
                Barcode.FORMAT_CODABAR -> CODABAR
                Barcode.FORMAT_EAN_13 -> EAN_13
                Barcode.FORMAT_EAN_8 -> EAN_8
                Barcode.FORMAT_ITF -> ITF
                Barcode.FORMAT_UPC_E -> UPC_E
                Barcode.FORMAT_QR_CODE -> QR
                Barcode.FORMAT_PDF417 -> PDF_417
                Barcode.FORMAT_AZTEC -> AZTEC
                Barcode.FORMAT_DATA_MATRIX -> DATA_MATRIX
                else -> UNKNOWN
            }
    }
}
